### PR TITLE
Choosing algorithmus genre levl

### PR DESCRIPTION
**Pull request by niw5385**
mainly two functions added:
(1) getGenreWithLevel -> calculates the genre and the level for the next question, should be called before a new item is generated, the generated item should be of the calculated genre and level
(2) updateUserData -> after the evaluation of a question the user data has to be adjusted by the reached score (0%-100%) of the questions

furthermore:
- data structure for user adjusted two improve the choosing of the next genre/level therefor short term history added

### CODE REVIEW

File: src/lib/firebase/dataBaseLoadings.ts [hunk 1]
```diff
@@ -81,7 +111,7 @@ export const saveHistory = async (docName: string, db: any)=>{
             const history = data.logs.history;
             const historyLength = history.length;
             const sessions = history[historyLength-1].sessions
-            const begin= sessions[sessions.length-1].begin ;
+            const begin = sessions[sessions.length-1].begin;
             const date = new Date();
             const duration = Math.round(date.valueOf()/1000 ) - begin.seconds;
```
Comment (EduardvonBriesen) [inline src/lib/firebase/dataBaseLoadings.ts:116]: there is a `getSeconds()` method you can use

File: src/lib/firebase/dataBaseLoadings.ts [hunk 2]
```diff
@@ -1,5 +1,15 @@
...
+const MAX_HISTORY_LENGTH = 20;
+const MAX_HISTORY_GENRE_LENGTH = 5;
+const WEIGHT_QUESTIONS_OVERALL = 1;
+const WEIGHT_SCORE_OVERALL = 1.5;
+const WEIGHT_SCORE_HISTORY = 4;
+const WEIGHT_QUESTIONS_HISTORY = 3.5;
+
+
+/* Exports */
 export type UserData = {
```
Comment (EduardvonBriesen) [inline src/lib/firebase/dataBaseLoadings.ts:13]: I would move the type definition to a seperat file (e.g. see api type defs)

File: src/lib/firebase/dataBaseLoadings.ts [hunk 3]
```diff
@@ -360,4 +423,308 @@ export const initDataStructure = (name: string, email: string )=> {
...
+                [`progress.genres.${genre}.history.index_oldest_score`]: genre_history_scores_index
+        });
+    }catch(e){
+        console.error(e);
+    }
+
+} 
+
+/* Helper functions */
+type LevelData = {
```
Comment (EduardvonBriesen) [inline src/lib/firebase/dataBaseLoadings.ts:519]: You can use these type definitions in the User Data type

File: src/lib/firebase/dataBaseLoadings.ts [hunk 4]
```diff
@@ -360,4 +423,308 @@ export const initDataStructure = (name: string, email: string )=> {
...
+                    + 0.1*Math.random();
+        }
+        scores.push(score);   
+    }
+    return genres[scores.indexOf(Math.min(...scores))];
+}
+
+/**
+ * Function to determine the level of the next quetsions (of type genre)
+ * @returns {Levels}
```
Comment (EduardvonBriesen) [inline src/lib/firebase/dataBaseLoadings.ts:611]: Method actually returns a string

**Review by EduardvonBriesen — CHANGES_REQUESTED** [review]
Looks good so far, most of the comments I left aren't concerning functionality.

I didn't actually test the code.

If you have any sources on the algorithms you used, please add those as a comment in the source code

File: src/lib/firebase/dataBaseLoadings.ts (hunk 1 above)
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts]: I tried it, it isn't working

File: src/lib/firebase/dataBaseLoadings.ts (hunk 2 above)
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts:12]: done

File: src/lib/firebase/dataBaseLoadings.ts (hunk 3 above)
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts:528]: it is not exactly the same structure

File: src/lib/firebase/dataBaseLoadings.ts (hunk 4 above)
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts:623]: changed it

File: src/lib/firebase/dataBaseLoadings.ts [hunk 5]
```diff
@@ -360,4 +423,308 @@ export const initDataStructure = (name: string, email: string )=> {
...
+type GenreScores ={
+    classic: number;
+    rock: number;
+    pop: number;
+    jazz: number;
+    rap: number;
+    folk_music: number;
+}
+
+const newHistoryArrayElement = (date: Date )=> {
```
[the remark this answers is not in the document]
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts]: I changed it

File: src/lib/firebase/dataBaseLoadings.ts [hunk 6]
```diff
@@ -360,4 +423,308 @@ export const initDataStructure = (name: string, email: string )=> {
...
+    const docRef = doc(db, collectionsName, docName);
+    const docSnap = await getDoc(docRef);
+    if(!docSnap.exists()){
+        return;
+    }
+
+    const userData = docSnap.data() as UserData;
+    const genre: Genre = getGenreForItemtype(userData);
+
+    const genreLevelData:LevelData = userData.progress.genres[genre] as LevelData;
```
[the remark this answers is not in the document]
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts]: I will tried it

File: src/lib/firebase/dataBaseLoadings.ts [hunk 7]
```diff
@@ -360,4 +423,308 @@ export const initDataStructure = (name: string, email: string )=> {
...
+    const level:Levels = getLevelForGenre(genreLevelData) as Levels;
+    return {
+        genre: genre, 
+        level: level
+    };
+}
+
+export async function updateUserProgressData(db: any, docName: string, relativePoints: number, genre: Genre, level: Levels){
```
[the remark this answers is not in the document]
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts]: adjusted

File: src/lib/firebase/dataBaseLoadings.ts [hunk 8]
```diff
@@ -1,5 +1,15 @@
-import { doc, getDoc,updateDoc,type DocumentData, Timestamp } from "firebase/firestore";
+import { doc, getDoc,updateDoc,type DocumentData} from "firebase/firestore";
```
[the remark this answers is not in the document]
Comment (niw5385) [reply src/lib/firebase/dataBaseLoadings.ts]: done

**Review by EduardvonBriesen — APPROVED** [review]
Made some small adjustments myself, please look over them. Can be merged otherwise

**Pull request merged**